# Project Summarization
1. Total **9** pipelines/tools
   * exclude some tools I developed for some old spesific requirements
      * We do not need to spend time to maintain them

2. Four categories, inlcuding

 - (1) Illumina short reads (DNA & RNA) -> 3
   * Primary pipeline
   * Prepare Untrimmed Reads (Tongwu's group)
   * COVID project (include both primary and 2nd pipeline)
   
 - (2) Pacbio long reads (HiFi reads) -> 1
   * structure variation detection for whole genome sequencing

 - (3) RNA Pipeline -> 2
   * LIMS_RNA_Pipeline (From Mia)
   * RNA_Auto_Launcher
 
 - (4) Scientific Support -> 3
   * Compare NovoAlign and BWA for the alternative scaffold regions
   * cgr-telomeres (Cameron, Casey)
   * CheckMatchAlleleDefinitionTable (NP0522-001 (Ultrasound Tamoxifen Breast Study Pac Bio) Sequencing Analysis) 

 - (5) Other old tools 
   * Please check the related section below

## Illumina short reads (DNA & RNA)
### 1. Primary pipeline
1. Track new flowcells
2. Debug the running error
3. Fix bugs
   * e.g. sending warning email when something bad occured in cluster
4. Gitlab: http://10.133.130.114/wenluo/primary_analysis_pipeline


### 2. Prepare Untrimmed Reads
1. Keep adapter
2. Reuse some code from primary pipeline
3. Gitlab: http://10.133.130.114/lxwgcool/PrepareUntrimmedReads

### 3. COVID project
1. Handle USU data
2. Cover the essential logic in both primary pipeline and 2nd pipeline
3. Run in biowulf
4. Use S3 storage system
5. Prepare BAM and multiple different types of QC report
6. Github: https://github.com/NCI-CGR/IlluminaSequencingAnalysis/tree/main/CustomizedQC

## Pacbio long reads
### 1. Structure variation detection for whole genome sequencing
1. For pacbio hifi resads
2. Integrate multiple different callers 
3. Research investigation
4. Check other existing & similar pipeline 
5. This is a completely new pipleine.
6. Github: https://github.com/NCI-CGR/PacbioSequencingAnalysis/tree/main/SV_Detection

## RNA Pipeline
### 1. LIMS_RNA_Pipeline
1. For generating RNA QC report when primary pipeline be finished
2. Mia transfered it to me
3. Fix bugs
4. Track the running status of the new RNA flowcell
5. Add new features & modify the pipeline to make it be compatible with latest changes from LIMS.
6. Gitlab: http://10.133.130.114/lxwgcool/LIMS_RNA_Pipeline

### 2. RNA_Auto_Launcher
1. Handle the primary pipeline and LIMS_RNA_Pipeline automatically (e.g. auto run)
2. Send notification email to the related users. 
3. Gitlab: http://10.133.130.114/lxwgcool/RNA_Auto_Launcher

## Scientific Support
### 1. Compare NovoAlign and BWA for the alternative scaffold regions
1. This is the most important part to let people know how big impact if we transfer from NovoAlign to BWA
2. Please check the ticket below for the details of solution
   * https://cgr-bugz.nci.nih.gov/f/cases/29441/

### 2. cgr-telomeres (Cameron, Casey)
1: transfer from Cameron
2: Github: https://github.com/NCI-CGR/cgrtelomeres

### 3. CheckMatchAlleleDefinitionTable
1: This is from the requirement
   * (NP0522-001 (Ultrasound Tamoxifen Breast Study Pac Bio) Sequencing Analysis) 
2ï¼šGitlab
   * http://10.133.130.114/lxwgcool/CheckMatchAlleleDefinitionTable
3: Ticket
   * https://cgr-bugz.nci.nih.gov/f/cases/24320/

## Other tools (no need maintenance)
1. Github & Gitlab
   * http://10.133.130.114/lxwgcool/CommonTools/tree/master
   * https://github.com/NCI-CGR/IlluminaSequencingAnalysis/tree/main/Tools/UCSC_Bed
   * https://github.com/NCI-CGR/IlluminaSequencingAnalysis/tree/main/VCFClassification
