```
1. Project: Primary pipeline: responsible for the upstream data analysis for all sequencing data generated by CGR.
(1) Collaborate with Herb to address the issue of "Repeat of BN6, AZ8, DK3 sample subset that failed library prep in NP0403-HE19" (FG 30073, solved). 
(2) Fix incorrect logics, such as the calculation of "ON_TARGET_READS", "% lane", and some metrics in the phase of dedup (solved).
(3) Fix some bugs, such as abnormal value in WES QC report (solved). 
(4) Analyze many WGS and WES flowcells.

2: Project: Pacbio SV Pipeline for Mila's project (PI: Mila, done)
(1) WGS of 4 cell lines-structural variant calling on Milaâ€™s samples (FG 30037)
(2) Use the different combination (2 Aligner, 3 SV caller, 2 criteria) to call some specific SVs that Mila interested. 
(3) Deliver results to Mila.
(4) Explain why such SV was ignored by some SV callers. 

3: Project: Pacbio SV Pipeline for Benchmark
(1) Benchmarking Variant Callers with Sequel II Data (GiaB, Standard and Ultra-low input libraries) (FG 3000)
(2) Use HG002 to check the performance from different combinations for SV detection (2 Aligner, 3 SV caller, 2 criteria) (done)
(3) Compile Truvari in CCAD cluster, debug the running error from Truvari and use it for Benchmark comparison. (done)
(4) Develop a script to summarize the running results from  Truvari and generate the summarization table. (done)
(5) Summarize the guideline to let PI know how to make the choice for different dataset and research purpose (TBD). 

4: Project: Pacbio SV pipeline for general use. 
(1) Develop a general use SV pipeline to process the in-house Pacbio HIFI sequencing data.
(2) Several steps, including Data Structure Normalization, Pre-QC, Alignment, SV Calling, Get consensus call set, Post-QC. 
(3) Finished the architecture of pipeline, implemented the first 4 steps.
(4) Try both NanoPlot and Filtlong for Pre-QC. 
(5) Implement both voting method and ML method to get consensus call set, and compare their performance by using different dataset. 
(6) Evaluate the boosting-based and bagging-based ML method.
(7) Discuss with Pacbio staff about the missing Info metrics in pbsv call set. 

5: Project RNA pipeline & Mentor
(1) RNA_Auto_Launcher: add a new functions into to protect the specific lims-drop-box folder.
(2) LIMS RNA pipleline: analyze multiple RNA flowcells. 
(3) Work transfer & Mentor (RNA pipeline)
     a) Help Chad understand "RNA_Auto_Launcher" and "LIMS RNA pipleline" from both the requirement-level and code-level
     b) Help Chad address the issues when he run pipeline in CCAD cluster
     c) Help Chad build testing folder and run the testing code in CCAD cluster.
     d) Help Chad add new functions into the RNA pipeline 
     e) Help Chad debug the cronjob and snakemake issue from RNA pipeline
     f) Help Chad address the issue of gitlab
     g)  Help Chad understand the mechanism of github, how to submit code, record modifications and the high quality way to track the commit history. 
(4) Help chad to run UMI in biowulf cluster, how to do parallel computing, and debug the RAM issues. 

6: Project: COVID Pipeline 
(1) Summarize the existing progress of COVID pipeline. rearrange the existing running result. Introduce how to use command line to check these results. Create online docs to record these info.
(2) Details, please check:
https://github.com/NCI-CGR/IlluminaSequencingAnalysis/blob/main/Note/Meeting/04_05_2022_COVID_Meeting.md

```
